L85-core-binder-transport.md

```diff
@@ -0,0 +1,181 @@
+protocol.
+
+[JNI] and Java will be used by the transport implementation to set up the
+connection (APIs for setting up connections are only available in Java).
+After we establish the connection between server and client, the rest can be
+done in pure C++.
+
+### Client side API
+
+Since the channel creation process requires interactions with Android runtime,
+we will have a special API for creating binder transport channels.
+
+```cpp
+void CreateBinderChannel(
+    void* jni_env, jobject context, const std::string& package_name,
+    const std::string& class_name,
+    std::shared_ptr<grpc::binder::SecurityPolicy> security_policy,
+    std::function<void(void*, std::shared_ptr<grpc::Channel>)>
+        on_channel_created);
+
+class SecurityPolicy {
+ public:
+  // return true if the uid is authorized to connect
+  virtual bool IsAuthorized(void* jni_env, int uid) = 0;
+};
+```
+
+`jni_env`: Pointer to [JNIEnv]. It is required because the transport call into Java code to
+establish the connection.
+
+`context`: The service will only be considered required by this client for as
+long as the `context` exists (See [bindService] for details).
+
+`package_name` and `class_name`: These 2 arguments will be used to create a
+[ComponentName], which identifies a specific service to connect to.
+
+`security_policy`: Instance of class that can check if a given uid is authorized
+to connect. We will provide some possible implementations of this class. Some
+examples are (a) Allows all connection. (b) Allows the connection if signature
+is the same. (c) Allows the connection if the other end's APK is signed by Google.
+
+`on_channel_created`: a callback function that will be invoked when the
+`grpc::Channel` became available. We need to use asynchronous callback to notify
+user because the operation of "bind to the server" is asynchronous in its
+nature. User should not block the thread to wait for this callback function to
+be called. See notes at the end of this RFC for details.
+* `JNIEnv` is provided so user can call their Java code if they would like.
+* `std::shared_ptr<grpc::Channel>` is the gRPC channel. It will be nullptr if the bind is not successful.
+
+The API is implemented using `[bindService]` in Java. Upon [onServiceConnected]
+called by Android, the channel will be created and the callback will be invoked.
+
+Because the API's signature contains [JNI] specific type `jobject`, the API will
+only be declared if the header is compiled with Android tool chain. (that is,
+when `GPR_ANDROID` is defined)
+
+### Server side API
+
+Service implementation will be an Android [bound service] in Java, and it should return
+a binder object created by our transport implementation to the client.
+
+We propose to introduce a new API that returns credentials for binder server:
+
+```Cpp
+std::shared_ptr<ServerCredentials> BinderServerCredentials(std::shared_ptr<grpc::binder::SecurityPolicy> security_policy);
```

Comment:
We should include the C API `grpc_server_add_binder_port` into this RFC per discussions made in https://github.com/grpc/grpc/pull/27036.

Reply:
Mark and Craig: Can you confirm if we only need C API for this specific API are we need C API for every C++ API? What's the reason to introduce C API?